pkg/metrics/metrics.go

```diff
@@ -0,0 +1,145 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/keep-common/pkg/metrics"
+	"github.com/keep-network/keep-core/pkg/chain"
+	"github.com/keep-network/keep-core/pkg/net"
+)
+
+var logger = log.Logger("keep-metrics")
+
+// Initialize set up the metrics registry and enables metrics server.
+func Initialize(
+	port int,
+) (*metrics.Registry, error) {
+	if port == 0 {
+		return nil, fmt.Errorf("invalid port")
+	}
+
+	registry := metrics.NewRegistry()
+
+	registry.EnableServer(port)
+
+	return registry, nil
+}
+
+// ObserveConnectedPeersCount triggers an observation process of the
+// connected_peers_count metric.
+func ObserveConnectedPeersCount(
+	ctx context.Context,
+	registry *metrics.Registry,
+	netProvider net.Provider,
+	tick time.Duration,
+) {
+	input := func() float64 {
+		connectedPeers := netProvider.ConnectionManager().ConnectedPeers()
+		return float64(len(connectedPeers))
+	}
+
+	observe(
+		ctx,
+		"connected_peers_count",
+		input,
+		registry,
+		tick,
+	)
+}
+
+// ObserveConnectedBootstrapPercentage triggers an observation process of the
+// connected_bootstrap_percentage metric.
+func ObserveConnectedBootstrapPercentage(
```

Comment:
This is a neat metric - 

What stands out is that an operator must know how many peers are in their `Peers` list to reasonably assert the meaning of the value reported here.  

Simple example:  Person A  has 2 peers in their list and they can't reach one of them, this metric will report 50% bootstraps connected.  Person B has 20 peers in their list and you can't reach half of them this metric will report 50% bootstraps connected. It's reasonable to say In this scenario person A should take action to mitigate loosing the network in case they reboot, whereas person B can still sip their coffee in comfort.

It's reasonable to assume that an operator **should** know what their configurations contain - but I could see this being a possible point of confusion for "actionable data".  Maybe "actionable data" isn't the point here?

Digging a bit closer to what I'm feeling here, **what** is it that we want to surface with these metrics (the scope)?  It could be as simple as "surface an operators state in relation to its configuration and current state of the network".  What is the intent of surfacing these metrics?

From an operator perspective I'm of the opinion that each metric surfaced should represent a potentially actionable state.  We should take care to articulate that that state in the description for each metric.  "actionable" is highly opinionated however using this metric as an example we can say "If this reads 0, you cannot connect to the network if your machine restarts"

Once we make these statements it's easier to see if there is overlap between metrics, etc.

To be clear, I'm 100% ok leaving this as-is, mostly wanting to flesh out the motivation.

Reply:
Regarding those `Connected*` metrics my intentions were as follows:
- `ConnectedPeersCount`: I realized we often check this parameter in our logs treating it as a general clue about the network health and the magnitude of possible load. Apart from that, if something goes wrong with the node, this parameter often falls suddenly. Hence, I think this is a good candidate for monitoring.
- `ConnectedBootstrapPercentage`:  This one is here to give a quick response to the question: "what's the condition of my configured network entry point?" and raise an alert if the condition is rather bad.  If someone configures only two bootstraps, this is their own risk. This metric just shows the situation relative to the `LibP2P.Peers` property. I think we must be relative here because we can't easily say how much bootstraps on the list are enough as it depends of the network size.  Second thing, we use `percentage` instead of `count` because I think we can have some troubles when defining alerts if we use the latter. For example, if we decide to alert when the bootstrap `count` is less than `5`, it may be a true alert for a node with`100` configured peers but not quite for a node with `10` configured peers. Defining an alert using `percentage` is easier and means the same for every node.

Reply:
> "what's the condition of my configured network entry point?" 

I like this - and did not position it that way when working through the code initially.

Reply:
> Second thing, we use percentage instead of count because I think we can have some troubles when defining alerts if we use the latter. For example, if we decide to alert when the bootstrap count is less than 5, it may be a true alert for a node with100 configured peers but not quite for a node with 10 configured peers. Defining an alert using percentage is easier and means the same for every node.

I do agree with everything else but I don't think I agree with this single statement. 😆 

If we want to answer the question of "what's the condition of my configured entry point", I think we should be very explicit and expect the operator to define their comfort level themselves. Just like @sthompson22's noted, 50% of 4 is something else than 50% of 200.

I like we have a separate metric for all connected peers and for bootstrap peers but I do think bootstrap peers metric should be a number, not a percentage and the operator should define the alert knowing their configuration and number of bootstrap peers defined there.

Reply:
Ok, we can change `connected_bootstrap_percentage` to `connected_bootstrap_count`. Having it as percents seemed handier at the beginning but after summarizing all points, making it as explicit count appears as a better option now.